method/CellCharter/CellCharter.py

```diff
@@ -0,0 +1,198 @@
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, 
+                       var=features, 
+                       obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = sp.io.mmread(args.matrix)
+        if sp.sparse.issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+## Your code goes here
+import cellcharter as cc
+import anndata as ad
+import scanpy as sc
+import squidpy as sq
+import pandas as pd
+import numpy as np
+import random
+
+random.seed(seed)
+
+# Load data
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+# raw count input
+import scvi
+
+# Assume filtered or not?
+# sc.pp.filter_genes(adata, min_counts=3)
+#sc.pp.filter_cells(adata, min_counts=3)
+
+# preprocessing for scvi
+adata.layers["counts"] = adata.X.copy()
+
+# Set up scvi model for reduced_dimension embedding
+scvi.settings.seed = seed
+scvi.model.SCVI.setup_anndata(
+    adata,
+    layer="counts"
+)
+
+model = scvi.model.SCVI(adata)
+model.train(early_stopping=True, enable_progress_bar=True)
+adata.obsm['reduced_dimensions'] = model.get_latent_representation(adata).astype(np.float32)
+
+# Find neighbors
+coord = "grid" if technology in ["Visium", "ST"] else "generic"
+delTri = technology not in ["Visium", "ST"] 
+
+sq.gr.spatial_neighbors(adata, 
+                        coord_type = coord, 
+                        delaunay=delTri)
+
+# Trim if only delaunay is used
+if delTri:
+    cc.gr.remove_long_links(adata)
+
+# Aggregate neighborhoods
+cc.gr.aggregate_neighbors(adata, 
+                          n_layers=config["n_layers"], 
+                          aggregations=config["aggregations"],
+                          use_rep='reduced_dimensions', 
+                          out_key='X_cellcharter', 
+                          sample_key=None)
+
+# clustering
+model = cc.tl.Cluster(
+    n_clusters=n_clusters, 
+    random_state=seed,
+    convergence_tolerance=config["convergence_tolerance"], 
+    covariance_regularization=config["covariance_regularization"]
+    # if running on gpu
+    #trainer_params=dict(accelerator='gpu', devices=1)
```

Comment:
So can we currently use GPU for this tool?

Reply:
Yes -- I added a cuda of GPU availability now.